core/src/smtp/hotmail.rs

```diff
@@ -140,9 +151,65 @@ pub async fn check_password_recovery(
 	})
 }
 
+/// Convert an email address to its corresponding OneDrive URL.
+fn get_onedrive_url(email_address: &str) -> String {
+	let (username, domain) = email_address
+		.split_once('@')
+		.expect("Email address syntax already validated.");
+	let (tenant, _) = domain
+		.split_once('.')
+		.expect("Email domain syntax already validated.");
+
+	format!(
+		"https://{}-my.sharepoint.com/personal/{}_{}/_layouts/15/onedrive.aspx",
+		tenant,
+		username.replace('.', "_"),
+		domain.replace('.', "_"),
+	)
+}
+
+/// Use a HTTP request to verify if an Microsoft 365 email address exists.
+///
+/// See
+/// [this article](<https://www.trustedsec.com/blog/achieving-passive-user-enumeration-with-onedrive/>)
+/// for details on the underlying principles.
+///
+/// Note that a positive response from this function is (at present) considered
+/// a reliable indicator that an email-address is valid. However, a negative
+/// response is ambigious: the email address may or may not be valid but this
+/// cannot be determined by the method outlined here.
+pub async fn check_microsoft365_api(
+	to_email: &EmailAddress,
+	input: &CheckEmailInput,
+) -> Result<Option<SmtpDetails>, HotmailError> {
```

Comment:
I just tried this manually, and I believe there's a bug: if you run `cargo run -- --microsoft365-use-api=true <email>` with someone that hasn't opened their one drive, it returns a ReqwestError. Whereas I believe in https://github.com/reacherhq/check-if-email-exists/pull/1194#issuecomment-1271578587 we both meant that in that case the algorithm should just continue as normal.

Reply:
I'm inclined to simply do this:

```suggestion
) -> Result<SmtpDetails, HotmailError> {
```

and  below do:

```rust
	if input.microsoft365_use_api && host_lowercase.ends_with(".mail.protection.outlook.com.") {
		// Continue in the event of an ambiguous result.
		if let Ok(smtp_details) = hotmail::check_microsoft365_api(to_email, input).await {
			return Ok(smtp_details);
		}
	}
```

Reply:
Ah, I think I know what's happening:

1. if the API returns a 403, it returns a positive response (`Ok(Some(smtp_details))`.)
1. if the API returns a 404, it returns a negative response (`Ok(None)`.)
1. if the API fails because the OneDrive domain doesn't exist (i.e. the company in question just doesn't use OneDrive), it raises an error (`Err(HotmailError(ReqwestError))`.)
1. if the API fails for some other reason (e.g. a transient network issue), it raises an error (`Err(HotmailError(ReqwestError))`.)

The problem is that those last two are indistinguishable, although 3. should result in some indication that it's ambiguous, like 1. and 2.

However, I can't of a way to make that distinction so I think you're right: I'll aim to allow it continue in the event of _any_ error (but might add an additional log which might allow things to be narrowed down in the future.)

Reply:
@amaurym, I've pushed up a change which resolves the issues (I think; I've tested the scenarios I'm aware of any they either pass or continue to SMTP checks) but leaves the `Result<Option<SmtpDetails>, HotmailError>` signature in place.

In order to have `check_microsoft365_api()` return a `Result<SmtpDetails, HotmailError>`, I'd have to change the "successful" 404 response into a `HotmailError`. That's easy enough but adding another `HotmailError` type (e.g. `HotmailError::AmbiguousResponse`) didn't seem quite right.

Basically:

- if the HTTP request succeeds and we see a 403, we return the result.
- if the HTTP request succeeds and we see a 404, we continue to SMTP checks.
- if the HTTP request fails, we log the error and continue to SMTP checks.